Add blog post about OCI security profiles (#40485)
kubernetes/website: content/en/blog/_posts/2023-04-05-oci-security-profiles.md

seccomp profiles which are way simpler and scoped to the application logic,
without having a need to take the whole infrastructure setup into account.

[runc]: https://github.com/opencontainers/runc
[crun]: https://github.com/containers/crun

But how to maintain those base profiles? For example, the amount of required
syscalls for a runtime can change over its release cycle in the same way it can
change for the main application. Base profiles have to be available in the same
cluster, otherwise the main seccomp profile will fail to deploy. This means that
they’re tightly coupled to the main application profiles, which acts against the
main idea of base profiles. Distributing and managing them as plain files feels
like an additional burden we have to solve.

## OCI artifacts to the rescue

The [latest release][spo-latest] of the Security Profiles Operator supports
managing base profiles as OCI artifacts! Imagine OCI artifacts as lightweight
container images, storing files in layers in the same way images do, but without
a process to be executed. Those artifacts can be used to store security profiles
like regular container images in compatible registries. This means they can be

> versionized, namespaced and annotated similar to regular container images.

Reviewer: ```suggestion
versioned, namespaced and annotated similar to regular container images.
```

[spo-latest]: https://github.com/kubernetes-sigs/security-profiles-operator/releases/latest

> To see how that works in action, we have to specify a `baseProfileName` prefixed

Reviewer 2: ```suggestion
To see how that works in action,  specify a `baseProfileName` prefixed
```
We [avoid using we](https://kubernetes.io/docs/contribute/style/style-guide/#avoid-using-we) in tech docs and in other contexts where it's not clear who the “we” refers to.

with `oci://` within a seccomp profile CRD, for example:

```yaml
apiVersion: security-profiles-operator.x-k8s.io/v1beta1
kind: SeccompProfile
metadata:
  name: test
spec:
  defaultAction: SCMP_ACT_ERRNO
  baseProfileName: oci://ghcr.io/security-profiles/runc:v1.1.5
  syscalls:
    - action: SCMP_ACT_ALLOW
      names:
        - uname
```

The operator will take care of pulling the content by using [oras][oras], as
well as verifying the [sigstore (cosign)][cosign] signatures of the artifact. If
the artifacts are not signed, then the SPO will reject them. The resulting
profile `test` will then contain all base syscalls from the remote `runc`
profile plus the additional allowed `uname` one. It is also possible to
reference the base profile by its digest (SHA256) making the artifact to be
pulled more specific, for example by referencing
`oci://ghcr.io/security-profiles/runc@sha256:380…`.

[oras]: https://oras.land
[cosign]: https://github.com/sigstore/cosign

The operator internally caches pulled artifacts up to 24 hours for 1000

> profiles, meaning that they will be refreshed after that time period, if the

Reviewer 3: Is it up to 24 hours and up to 1000 profiles? What happens when there's more than 1000 profiles?

Author: Yep, both limitations will apply. The oldest will be dropped and we will re-cache it.

cache is full or the operator daemon gets restarted.

> Because the overall resulting syscalls are hidden to the user (we just have the
> `baseProfileName` in the CRD, not the syscalls itself), we additionally annotate
> the seccomp profile with the final `syscalls`:

Reviewer 2: Try this:
```suggestion
Because the overall resulting syscalls are hidden from the user (I only have the
`baseProfileName` listed in the SeccompProfile, and not the syscalls themselves), I'll additionally
annotate that SeccompProfile with the final `syscalls`.

Here's how the SeccompProfile looks after I annotate it:
```

---

**Aside**
This is a private annotation - is this just a hint to the humans operating the workload?
If the annotation is intended to be machine readable / consumed elsewhere that's an architecture bug.
Kubernetes code should use public annotations (eg `spo.x-k8s.io/allowed-system-calls`).

Author: > This is a private annotation - is this just a hint to the humans operating the workload?

Yes, nothing machine readable.

```console
> kubectl describe seccompprofile test
Name:         test
Namespace:    security-profiles-operator
Labels:       spo.x-k8s.io/profile-id=SeccompProfile-test
Annotations:  syscalls:
                [{"names":["arch_prctl","brk","capget","capset","chdir","clone","close",...
API Version:  security-profiles-operator.x-k8s.io/v1beta1
```

The SPO maintainers provide all public base profiles as part of the [“Security
Profiles” GitHub organization][org].

[org]: https://github.com/orgs/security-profiles/packages

## Managing OCI security profiles

Alright, now the official SPO provides a bunch of base profiles, but how can I
define my own? Well, first of all we have to choose a working registry. There

> are a bunch of registries supporting OCI artifacts for now:

Reviewer: ```suggestion
are a bunch of registries that already supports OCI artifacts:
```

        This information will be used for signing this artifact and will be stored in public transparency logs and cannot be removed later.

By typing 'y', you attest that you grant (or have permission to grant) and agree to have this information stored permanently in transparency logs.
Your browser will now be opened to:
https://oauth2.sigstore.dev/auth/auth?access_type=…
Successfully verified SCT...
tlog entry created with index: 16520520
Pushing signature to: ghcr.io/security-profiles/crun
```

We can see that the tool automatically signs the artifact and pushes the
`./examples/baseprofile-crun.yaml` to the registry, which is then directly ready
for usage within the SPO. If username and password authentication is required,
either use the `--username`, `-u` flag or export the `USERNAME` environment
variable. To set the password, export the `PASSWORD` environment variable.

It is possible to add custom annotations to the security profile by using the
`--annotations` / `-a` flag multiple times in `KEY:VALUE` format. Those have no
effect for now, but at some later point additional features of the operator may
rely them.

> The `spoc` client is also able to pull security profiles from OCI artifact

Author: You can use the ORAS CLI for pulling the artifact: https://oras.land/cli

Docker does not support OCI artifacts right now. :)

compatible registries. To do that, just run `spoc pull`:

```console
> spoc pull ghcr.io/security-profiles/runc:v1.1.5
16:32:29.795597 Pulling profile from: ghcr.io/security-profiles/runc:v1.1.5
16:32:29.795610 Verifying signature

Verification for ghcr.io/security-profiles/runc:v1.1.5 --
The following checks were performed on each of these signatures:
  - Existence of the claims in the transparency log was verified offline
  - The code-signing certificate was verified using trusted certificate authority certificates

[{"critical":{"identity":{"docker-reference":"ghcr.io/security-profiles/runc"},…}}]
16:32:33.208695 Creating file store in: /tmp/pull-3199397214
16:32:33.208713 Verifying reference: ghcr.io/security-profiles/runc:v1.1.5
16:32:33.208718 Creating repository for ghcr.io/security-profiles/runc
16:32:33.208742 Using tag: v1.1.5
16:32:33.208743 Copying profile from repository
16:32:34.119652 Reading profile
16:32:34.119677 Trying to unmarshal seccomp profile
16:32:34.120114 Got SeccompProfile: runc-v1.1.5
16:32:34.120119 Saving profile in: /tmp/profile.yaml
```

> The profile can be now found in `/tmp/profile.yaml` or the specified output file

Reviewer 4: On all cluster nodes?

Author: No just on the local machine.
